File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/LocalVideosPresenter.java
package dev.ragnarok.fenrir.mvp.presenter;

import static dev.ragnarok.fenrir.util.Utils.isEmpty;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.db.Stores;
import dev.ragnarok.fenrir.model.LocalVideo;
import dev.ragnarok.fenrir.mvp.presenter.base.RxSupportPresenter;
import dev.ragnarok.fenrir.mvp.view.ILocalVideosView;
import dev.ragnarok.fenrir.util.Objects;
import dev.ragnarok.fenrir.util.RxUtils;
import dev.ragnarok.fenrir.util.Utils;


public class LocalVideosPresenter extends RxSupportPresenter<ILocalVideosView> {

    private final List<LocalVideo> mLocalVideos;
    private final List<LocalVideo> mLocalVideos_search;
    private boolean mLoadingNow;
    private String q;

    public LocalVideosPresenter(@Nullable Bundle savedInstanceState) {
        super(savedInstanceState);

        mLocalVideos = new ArrayList<>();
        mLocalVideos_search = new ArrayList<>();
        loadData();
    }

    private void loadData() {
        if (mLoadingNow) return;

        changeLoadingState(true);
        appendDisposable(Stores.getInstance()
                .localMedia()
                .getVideos()
                .compose(RxUtils.applySingleIOToMainSchedulers())
                .subscribe(this::onDataLoaded, this::onLoadError));
    }

    public void fireSearchRequestChanged(String q, boolean force) {
        String query = q == null ? null : q.trim();

        if (!force && Objects.safeEquals(query, this.q)) {
            return;
        }
        this.q = query;
        mLocalVideos_search.clear();
        if (!isEmpty(this.q)) {
            for (LocalVideo i : mLocalVideos) {
                if (isEmpty(i.getTitle())) {
                    continue;
                }
                if (i.getTitle().toLowerCase().contains(this.q.toLowerCase())) {
                    mLocalVideos_search.add(i);
                }
            }
        }

        if (!isEmpty(this.q))
            callView(v -> v.displayData(mLocalVideos_search));
        else
            callView(v -> v.displayData(mLocalVideos));
    }

    private void onLoadError(Throwable throwable) {
        changeLoadingState(false);
    }

    private void onDataLoaded(List<LocalVideo> data) {
        changeLoadingState(false);
        mLocalVideos.clear();
        mLocalVideos.addAll(data);
        resolveListData();
        resolveEmptyTextVisibility();
        if (!isEmpty(q)) {
            fireSearchRequestChanged(q, true);
        }
    }

    @Override
    public void onGuiCreated(@NonNull ILocalVideosView viewHost) {
        super.onGuiCreated(viewHost);
        resolveListData();
        resolveProgressView();
        resolveFabVisibility(false);
        resolveEmptyTextVisibility();
    }

    private void resolveEmptyTextVisibility() {
        callView(v -> v.setEmptyTextVisible(Utils.safeIsEmpty(mLocalVideos)));
    }

    private void resolveListData() {
        if (isEmpty(q)) {
            callView(v -> v.displayData(mLocalVideos));
        } else {
            callView(v -> v.displayData(mLocalVideos_search));
        }
    }

    private void changeLoadingState(boolean loading) {
        mLoadingNow = loading;
        resolveProgressView();
    }

    private void resolveProgressView() {
        callView(v -> v.displayProgress(mLoadingNow));
    }

    public void fireFabClick() {
        ArrayList<LocalVideo> localVideos = Utils.getSelected(mLocalVideos);
        if (!localVideos.isEmpty()) {
            callView(v -> v.returnResultToParent(localVideos));
        } else {
            callView(v -> v.showError(R.string.select_attachments));
        }
    }


    public void fireVideoClick(@NonNull LocalVideo video) {
        video.setSelected(!video.isSelected());

        if (video.isSelected()) {
            ArrayList<LocalVideo> single = new ArrayList<>(1);
            single.add(video);
            callView(v -> v.returnResultToParent(single));
        }
    }

    private void resolveFabVisibility(boolean anim) {
        resolveFabVisibility(Utils.countOfSelection(mLocalVideos) > 0, anim);
    }

    private void resolveFabVisibility(boolean visible, boolean anim) {
        callView(v -> v.setFabVisible(visible, anim));
    }

    public void fireRefresh() {
        loadData();
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/LocalVideosPresenter.kt
package dev.ragnarok.fenrir.mvp.presenter

import android.os.Bundle
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.db.Stores
import dev.ragnarok.fenrir.model.LocalVideo
import dev.ragnarok.fenrir.mvp.presenter.base.RxSupportPresenter
import dev.ragnarok.fenrir.mvp.view.ILocalVideosView
import dev.ragnarok.fenrir.util.Objects.safeEquals
import dev.ragnarok.fenrir.util.RxUtils.applySingleIOToMainSchedulers
import dev.ragnarok.fenrir.util.Utils.countOfSelection
import dev.ragnarok.fenrir.util.Utils.getSelected
import java.util.*

class LocalVideosPresenter(savedInstanceState: Bundle?) :
    RxSupportPresenter<ILocalVideosView>(savedInstanceState) {
    private val mLocalVideos: MutableList<LocalVideo>
    private val mLocalVideos_search: MutableList<LocalVideo>
    private var mLoadingNow = false
    private var q: String? = null
    private fun loadData() {
        if (mLoadingNow) return
        changeLoadingState(true)
        appendDisposable(Stores.instance
            .localMedia()
            .videos
            .compose(applySingleIOToMainSchedulers())
            .subscribe({ onDataLoaded(it) }) {
                onLoadError()
            })
    }

    fun fireSearchRequestChanged(q: String?, force: Boolean) {
        val query = q?.trim { it <= ' ' }
        if (!force && safeEquals(query, this.q)) {
            return
        }
        this.q = query
        mLocalVideos_search.clear()
        if (!this.q.isNullOrEmpty()) {
            for (i in mLocalVideos) {
                if (i.title.isNullOrEmpty()) {
                    continue
                }
                if (this.q?.lowercase(Locale.getDefault())
                        ?.let { i.title.lowercase(Locale.getDefault()).contains(it) } == true
                ) {
                    mLocalVideos_search.add(i)
                }
            }
        }
        if (!this.q.isNullOrEmpty()) view?.displayData(
            mLocalVideos_search
        )
        else view?.displayData(mLocalVideos)
    }

    private fun onLoadError() {
        changeLoadingState(false)
    }

    private fun onDataLoaded(data: List<LocalVideo>) {
        changeLoadingState(false)
        mLocalVideos.clear()
        mLocalVideos.addAll(data)
        resolveListData()
        resolveEmptyTextVisibility()
        if (!q.isNullOrEmpty()) {
            fireSearchRequestChanged(q, true)
        }
    }

    override fun onGuiCreated(viewHost: ILocalVideosView) {
        super.onGuiCreated(viewHost)
        resolveListData()
        resolveProgressView()
        resolveFabVisibility(false)
        resolveEmptyTextVisibility()
    }

    private fun resolveEmptyTextVisibility() {
        view?.setEmptyTextVisible(
            mLocalVideos
                .isNullOrEmpty()
        )
    }

    private fun resolveListData() {
        if (q.isNullOrEmpty()) {
            view?.displayData(
                mLocalVideos
            )
        } else {
            view?.displayData(
                mLocalVideos_search
            )
        }
    }

    private fun changeLoadingState(loading: Boolean) {
        mLoadingNow = loading
        resolveProgressView()
    }

    private fun resolveProgressView() {
        view?.displayProgress(mLoadingNow)
    }

    fun fireFabClick() {
        val localVideos = getSelected(mLocalVideos)
        if (localVideos.isNotEmpty()) {
            view?.returnResultToParent(
                localVideos
            )
        } else {
            view?.showError(R.string.select_attachments)
        }
    }

    fun fireVideoClick(video: LocalVideo) {
        video.isSelected = !video.isSelected
        if (video.isSelected) {
            val single = ArrayList<LocalVideo>(1)
            single.add(video)
            view?.returnResultToParent(
                single
            )
        }
    }

    private fun resolveFabVisibility(anim: Boolean) {
        resolveFabVisibility(countOfSelection(mLocalVideos) > 0, anim)
    }

    private fun resolveFabVisibility(visible: Boolean, anim: Boolean) {
        view?.setFabVisible(
            visible,
            anim
        )
    }

    fun fireRefresh() {
        loadData()
    }

    init {
        mLocalVideos = ArrayList()
        mLocalVideos_search = ArrayList()
        loadData()
    }
}